src/code/z_room.c

```diff
@@ -52,170 +52,207 @@ void Room_Draw0(PlayState* play, Room* room, u32 flags) {
+    MeshHeaderCull* meshHeaderCull;
+    MeshHeaderCullEntry* meshHeaderCullEntry;
+    MeshHeaderCullEntryLinked linkedEntriesBuffer[MESH_HEADER_CULL_MAX_ENTRIES];
+    MeshHeaderCullEntryLinked* head = NULL;
+    MeshHeaderCullEntryLinked* tail = NULL;
+    MeshHeaderCullEntryLinked* iter;
     s32 pad;
-    struct_80095D04* spA4;
+    MeshHeaderCullEntryLinked* insert;
     s32 j;
     s32 i;
     Vec3f pos;
     Vec3f projectedPos;
     f32 projectedW;
     s32 pad2;
-    MeshHeader2Entry* meshHeader2Entries;
-    MeshHeader2Entry* meshHeader2EntryIter;
-    f32 temp_f2;
+    MeshHeaderCullEntry* meshHeaderCullEntries;
+    MeshHeaderCullEntry* meshHeaderCullEntryIter;
+    f32 entryStartZ;
 
     OPEN_DISPS(play->state.gfxCtx, "../z_room.c", 287);
+
     if (flags & ROOM_DRAW_OPA) {
         func_800342EC(&D_801270A0, play);
         gSPSegment(POLY_OPA_DISP++, 0x03, room->segment);
         func_80093C80(play);
         gSPMatrix(POLY_OPA_DISP++, &gMtxClear, G_MTX_MODELVIEW | G_MTX_LOAD);
     }
+
     if (1) {}
+
     if (flags & ROOM_DRAW_XLU) {
         func_8003435C(&D_801270A0, play);
         gSPSegment(POLY_XLU_DISP++, 0x03, room->segment);
         Gfx_SetupDL_25Xlu(play->state.gfxCtx);
         gSPMatrix(POLY_XLU_DISP++, &gMtxClear, G_MTX_MODELVIEW | G_MTX_LOAD);
     }
 
-    meshHeader2 = &room->meshHeader->meshHeader2;
-    meshHeader2Entry = SEGMENTED_TO_VIRTUAL(meshHeader2->entries);
-    spA4 = spB8;
+    meshHeaderCull = &room->meshHeader->cull;
+    meshHeaderCullEntry = SEGMENTED_TO_VIRTUAL(meshHeaderCull->entries);
+    insert = linkedEntriesBuffer;
+
+    ASSERT(meshHeaderCull->numEntries <= ARRAY_COUNT(linkedEntriesBuffer), "polygon2->num <= SHAPE_SORT_MAX",
+           "../z_room.c", 317);
 
-    ASSERT(meshHeader2->numEntries <= SHAPE_SORT_MAX, "polygon2->num <= SHAPE_SORT_MAX", "../z_room.c", 317);
-    meshHeader2Entries = meshHeader2Entry;
+    meshHeaderCullEntries = meshHeaderCullEntry;
 
-    for (i = 0; i < meshHeader2->numEntries; i++, meshHeader2Entry++) {
-        pos.x = meshHeader2Entry->pos.x;
-        pos.y = meshHeader2Entry->pos.y;
-        pos.z = meshHeader2Entry->pos.z;
+    // Pick and sort entries by depth
+    for (i = 0; i < meshHeaderCull->numEntries; i++, meshHeaderCullEntry++) {
+
+        // Project the entry position, to get the depth it is at.
+        pos.x = meshHeaderCullEntry->pos.x;
+        pos.y = meshHeaderCullEntry->pos.y;
+        pos.z = meshHeaderCullEntry->pos.z;
         SkinMatrix_Vec3fMtxFMultXYZW(&play->viewProjectionMtxF, &pos, &projectedPos, &projectedW);
-        if (-(f32)meshHeader2Entry->unk_06 < projectedPos.z) {
-            temp_f2 = projectedPos.z - meshHeader2Entry->unk_06;
-            if (temp_f2 < play->lightCtx.fogFar) {
-                spA4->unk_00 = meshHeader2Entry;
-                spA4->unk_04 = temp_f2;
-                iter = spB4;
+
+        // If the entry isn't fully before the rendered depth range
```

Comment:
I'd say we could sphere it up a bit more here too, maybe say "entry's bounding sphere" or something like that. Could even introduce "unculling sphere" as a term of art, I suppose, although we've struggled with that style of terminology previously. Maybe just "sphere of visibility", lol. (Essentially the idea is to distinguish the simple geometric calculational idea from the actual drawing of the complex shape within.)